### DISCUSSION TITLE
Handle RediectUrl with PaymentGateway such as Paypal

### DISCUSSION THREAD

**Discussion by sonlm** [saleor/saleor#7184] [Deployment]
I am implementing paypal gateway. Everything works well except the checkOutComplete process. I added a returnUrl into Paypal API when creating Paypal Order. After user confirms the payment on Paypal site, it redirects user to the returnUrl that I have added before. At present, order is not created, I have to create the order after user confirms the payment. How can I show the order detail to user while it's not created yet ? And How can I handle the redirect url here ? The checkoutComplete precess does not handle this problem.

**Comment by Reef3rm4n** [comment]
@sonlm im also implementing paypal gateway care to share your progress so far ? I had a look this morning and it would seem that because the capture is done in paypal checkout saleor wont create the order.

**Reply by Reef3rm4n** [reply]
Is there a way/option for paypal to call the plugins webhook ? you could finish the order off when the hook gets called

**Reply by JannikZed** [reply]
is there a reason, why you don't want to use the existing Braintree Gateway?

**Reply by Reef3rm4n** [reply]
Hi @JannikZed Correct me if im wrong but it seems that braintree gateway does not implement the option to login and pay with user's paypal account but just credit card....

**Reply by JannikZed** [reply]
hey @JayBenato Braintree and PayPal works as expected:
![Screenshot 2021-05-05 at 09 17 39](https://user-images.githubusercontent.com/7526174/117107950-c527b680-ad82-11eb-8270-9390ea1e6e00.png)

we are using the Braintree DropIn UI.

**Reply by Reef3rm4n** [reply]
Hey @JannikZed,
awesome ! I will have a go at it today, did not know that it was working.
thanks

**Reply by Reef3rm4n** [reply]
@JannikZed do you mind sharing the code you've added on the front-end ? did you replace the CreditCardForm that the braintreePaymentGateway is returning ?

**Reply by JannikZed** [reply]
I think we used pretty much the standard dropIn UI code: 

```
 <DropIn
                            ref={dropinRef}
                            options={{
                                authorization: braintreeClientToken,
                                ...(checkout?.totalPrice?.gross?.amount > THREE_D_SECURE_THRESHOLD
                                    ? { threeDSecure: { amount: checkout?.totalPrice?.gross?.amount.toString() } }
                                    : {}
                                ),
                                // threeDSecure: { amount:  },
                                paypal: {
                                    // flow: 'vault',
                                    // amount: checkout?.totalPrice?.gross?.amount,
                                    // currency: 'EUR',

                                    flow: 'checkout',
                                    amount: checkout?.totalPrice?.gross?.amount,
                                    currency: 'EUR',
                                    commit: false,
                                },
                                // card: { cardholderName: { required: true } },
                                applePay: {
                                    // TODO: add applePay
                                    displayName: 'Pfeffer & Frost',
                                    paymentRequest: {
                                        // https://developer.apple.com/documentation/apple_pay_on_the_web/applepaypaymentrequest
                                        total: {
                                            // label: tCommon('cartInner.Total'),
                                            label: 'Pfeffer & Frost GmbH',
                                            amount: roundToFixed(checkout?.totalPrice?.gross?.amount),
                                            // type: 'pending',
                                        },
                                        countryCode: 'DE',
                                        currencyCode: 'EUR',
                                        shippingType: 'shipping',
                                        supportedNetworks: ["amex","cartesBancaires","chinaUnionPay","discover","eftpos","electron","elo","interac","jcb","mada","maestro","masterCard","privateLabel","visa","vPay"],
                                        merchantCapabilities: ["supports3DS", "supportsCredit", "supportsDebit"],
                                        // lineItems: [
                                        //     {
                                        //         label: tCommon('cartInner.subTotal'),
                                        //      amount: roundToFixed(checkout?.totalPrice?.gross?.amount - (checkout?.shippingMethod?.price?.amount || 0)),
                                        //         type: 'final',
                                        //     },
                                        //     {
                                        //         label: checkout?.shippingMethod?.name,
                                        //         amount: roundToFixed(checkout?.shippingMethod?.price?.amount),
                                        //         type: 'final',
                                        //     },
                                        // ],
                                    },
                                },
                                googlePay: {
                                    merchantId: process.env.NEXT_PUBLIC_GPAY_MERCHANT_ID || '',
                                    transactionInfo: { // https://developers.google.com/pay/api/web/reference/request-objects#TransactionInfo
                                        currencyCode: 'EUR',
                                        countryCode: 'DE',
                                        transactionId: checkout?.id,
                                        totalPriceStatus: 'FINAL',
                                        totalPrice: roundToFixed(checkout?.totalPrice?.gross?.amount),
                                        // checkoutOption: 'DEFAULT',
                                    },
                                },
                                locale: lang === 'de' ? 'de_DE' : 'en_US',
                            }}
                            onInstance={(localInstance) => {
                                // This will be true if you generated the client token
                                // with a customer ID and there is a saved payment method
                                // available to tokenize with that customer.
                                if (localInstance.isPaymentMethodRequestable()) {
                                    setMethodAvailable(true);
                                }
                                setInstance(localInstance);
                            }}
                        />
```

**Reply by tmechsner** [reply]
@JannikZed I think braintree does not provide react components. Did you use one of the third party react wrappers or did I miss something? Did you have to make further adjustments to the backend or is Paypal supported now by the saleor braintree gateway implementation? I saw that you closed your issues regarding this topic and found some merged pull requests.. thanks in advance :)

**Reply by JannikZed** [reply]
We are using this component: https://www.npmjs.com/package/braintree-web-drop-in-react
It is a React-Wrapper around the dropinUI from Braintree. Saleor does support Paypal, Google Pay and Apple Pay as well with the braintree payment gateway. - there was one issue, that got fixed already and is integrated in the 3.0 Saleor.

**Reply by tmechsner** [reply]
Thanks a lot for the quick reply! I looked at that component already, but it does not seem to be well maintained - last commit 11 months ago, does not support latest version of braintree dropinUI and no reaction on issues... I'll have a look at the forks or create an own one I think. Thanks again! :)

**Reply by JannikZed** [reply]
keep us updated, what you are doing! We are interested in a well maintained react library as well :P

**Reply by tmechsner** [reply]
Sure. :) Will start with that in two weeks, after my vacation.

**Reply by tmechsner** [reply]
@JannikZed  I didn't start an own react component, but instead used braintree in combination with the paypal sdk. I have two problems: Often after entering payment information, I get redirected to the review page and then back to shipping immediately. Did you coma cross that too? Also, I often encounter problems with Paypal: It says that it blocked the payment because of duplicate payment. This happens if a payment is created multiple times with the same order id. Any thoughts on that?

**Reply by JannikZed** [reply]
well - the dropin-ui works good for us, so we never came across that issues. So I can't help with that unfortunately.

**Comment by jor23dan** [comment]
can you  tell me  what to do implementing paypal gateway?

**Comment by jor23dan** [comment]
> I am implementing paypal gateway. Everything works well except the checkOutComplete process. I added a returnUrl into Paypal API when creating Paypal Order. After user confirms the payment on Paypal site, it redirects user to the returnUrl that I have added before. At present, order is not created, I have to create the order after user confirms the payment. How can I show the order detail to user while it's not created yet ? And How can I handle the redirect url here ? The checkoutComplete precess does not handle this problem.

can you tell me what to do implementing paypal gateway?
